tlatools/org.lamport.tlatools/test/tlc2/debug/ExpressionBreakpointTest.java

```diff
@@ -0,0 +1,66 @@
+package tlc2.debug;
+
+import org.eclipse.lsp4j.debug.SetBreakpointsArguments;
+import org.eclipse.lsp4j.debug.SourceBreakpoint;
+import org.junit.Test;
+import org.junit.Assert;
+
+import tlc2.output.EC;
+import tlc2.value.impl.IntValue;
+import tlc2.value.impl.LazyValue;
+import util.UniqueString;
+
+public class ExpressionBreakpointTest extends TLCDebuggerTestCase {
+
+	private static final String FOLDER = "debug";
+	private static final String RM = "ExpressionBreakpointTest";
+
+	public ExpressionBreakpointTest() {
+		super(RM, FOLDER, new String[] { "-config", RM + ".tla" }, EC.ExitStatus.SUCCESS);
+	}
+	
+	private static int getContextVal(TLCStackFrame frame, String var) {
+		Object result = frame.getContext().lookup(s -> s.getName().equals(var));
+		if (result instanceof IntValue) {
+			return ((IntValue)result).val;
+		} else if (result instanceof LazyValue) {
+			// Evade cache guard by guessing value from small range.
+			LazyValue lazy = (LazyValue)result;
+			for (int i = 0; i < 20; i++) {
+				if (lazy.equals(IntValue.gen(i))) {
+					return i;
+				}
+			}
+			throw new IllegalArgumentException();
+		} else {
+			throw new IllegalArgumentException();
+		}
+	}
+	
+	private static int getStateVal(TLCStackFrame frame, String var) {
+		return ((IntValue)frame.getT().getVals().get(UniqueString.of(var))).val;
+	}
+
+	@Test
+	public void testSpec() throws Exception {
+		// Set breakpoint
+		SourceBreakpoint bp = createBreakpointInfo(8, 5, 1, "(i + l + k) > j");
+		SetBreakpointsArguments sba = createBreakpointArgument(RM, bp);
+		debugger.setBreakpoints(sba);
+
+		// Break at init
+		debugger.continue_();
+		
+		// Break at bp
```

Comment:
This block can be simplified into:

```java
// Break at bp
final TLCStateStackFrame current = (TLCStateStackFrame) debugger.continue_()[0];
assertEquals(2, ((IntValue) current.getT().getVals().get(UniqueString.of("i"))).val);
final Map<String, String> allVariables = new HashMap<>();
allVariables.put("k", "5");
allVariables.put("l", "4");
assertTLCStateFrame(current, 8, 8, RM, allVariables);
```

`getStateVal` and `getContextVal` become obsolete.

Reply:
I kind of like the test as it is. It's a bit more clear that a property is being tested instead of hardcoding specific expected values.

Reply:
For consistency, please stick to the style that is used by the other `tlc2.debug` tests.

Reply:
I will change it, but I would like to lodge feedback that I find these type of tests confusing since it is never clear if they're just recording & playing back behavior or actually testing a property that matters. A good example is the test failure that was caused by changing the `pathTo` function, which manifested in a very weird way.